sycl/doc/extensions/proposed/sycl_ext_oneapi_reduction_properties.asciidoc

```diff
@@ -0,0 +1,238 @@
+compile-time properties to a reduction object.
+
+Each new overload has the same behavior as its corresponding definition in the
+SYCL 2020 specification unless the definition of a property passed in via the
+final `sycl::ext::oneapi::experimental::properties` parameter says otherwise.
+
+[source,c++]
+----
+namespace sycl {
+
+template <typename BufferT, typename BinaryOperation, typename PropertyList>
+__unspecified__ reduction(BufferT vars, handler& cgh, BinaryOperation combiner,
+                          PropertyList properties);
+
+template <typename T, typename BinaryOperation, typename PropertyList>
+__unspecified__ reduction(T* var, BinaryOperation combiner,
+                          PropertyList properties);
+
+template <typename T, typename Extent, typename BinaryOperation, typename PropertyList>
+__unspecified__ reduction(span<T, Extent> vars, BinaryOperation combiner,
+                          PropertyList properties);
+
+template <typename BufferT, typename BinaryOperation, typename PropertyList>
+__unspecified__
+reduction(BufferT vars, handler& cgh, const BufferT::value_type& identity,
+          BinaryOperation combiner, PropertyList properties);
+
+template <typename T, typename BinaryOperation, typename PropertyList>
+__unspecified__ reduction(T* var, const T& identity, BinaryOperation combiner,
+                          PropertyList properties);
+
+template <typename T, typename Extent, typename BinaryOperation, typename PropertyList>
+__unspecified__ reduction(span<T, Extent> vars, const T& identity,
+                          BinaryOperation combiner,
+                          PropertyList properties);
+
+}
+----
+
+=== Reduction properties
+
+New `reduction` properties are introduced to allow developers to constrain
+reduction algorithm selection based on desired behavior(s). Compile-time
+properties corresponding to existing runtime properties are also introduced to
+ensure that all information can be passed via a single property list.
+
+If a reduction kernel is submitted to a device that cannot satisfy the
+request for specific reduction behavior(s), the implementation must throw an
+`exception` with the `errc::feature_not_supported` error code.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+struct deterministic_key {
+  using value_t = property_value<deterministic_key>;
+};
+inline constexpr deterministic_key::value_t deterministic;
+
+struct initialize_to_identity_key {
+  using value_t = property_value<initialize_to_identity_key>;
+};
+inline constexpr initialize_to_identity_key::value_t initialize_to_identity;
+
+}
+----
+
+|===
+|Property|Description
+
+|`deterministic`
+|Adds the requirement that any two calls to the reduction on the same device
+which use the same launch configuration (i.e., `range` or `nd_range`) and in
+which the same kernel instances contribute the same values to the reduction
+will produce results with the same value.
```

Comment:
I understand the intent now, but I still think this wording is confusing.  In particular, the phrase "in the same kernel instances" throws me off.  Could we describe it like this?

> When two reductions both have this property, they are guaranteed to produce the same result when all of the following conditions hold:
> * Both calls run on the same device;
> * Both reductions are either in the same kernel invocation or are in kernels invoked with the same launch configuration (i.e. `range` or `nd_range`); and
> * The same values are contributed to each reduction.

Reply:
I'm happy to reformat it as a bulleted list, but the middle bullet isn't what I meant by "same kernel instances".

What I basically mean is that the same work-items have to be responsible for contributing the same values.  I was trying to avoid saying "work-item" because `reduction` can be used with both `range` and `nd_range` kernels... I think the easiest way to settle on wording here might be to look at some examples.

**Example 1**
```c++
q.parallel_for(N, sycl::reduction(...), [=](size_t i, auto& reducer) {
  reducer += data[i];
});
```
Every time this kernel is run, iteration `i` contributes `data[i]` to the reduction.  Each invocation of the kernel lambda with a different value of `i` is what I referred to as "a kernel instance".

**Example 2**
```c++
q.parallel_for(sycl::nd_range<1>(G, L), sycl::reduction(...), [=](sycl::nd_item<1> it, auto& reducer) {
  reducer += data[it.get_global_id(0)];
});
```
Every time this kernel is run, the value contributed by `it` to the reduction is invariant because it's derived from it's global index. I do think "each work item" would describe this case.

**Example 3**
```c++
q.parallel_for(sycl::nd_range<1>(G, L), sycl::reduction(...), [=](sycl::nd_item<1> it, auto& reducer) {
  size_t index = sycl::atomic_ref<...>(counter)++;
  reducer += data[index];
});
```

This is a case that *wouldn't* be guaranteed to produce deterministic results, because the assignment of values to work-items itself is non-deterministic.  This is the case I'm struggling to describe.  I wanted to draw attention to such cases to help people understand that adding `deterministic` won't have the desired effect.

Reply:
Oh, I see.  This distinction was not clear to me in the current wording.  I think the SYCL spec *does* use the term "work-item" even for a basic "range" parallel-for.  For example see [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_basic_kernels).  Could we describe it like this:

> When two reductions both have this property, they are guaranteed to produce the same result when all of the following conditions hold:
> * Both calls run on the same device;
> * Both reductions are either in the same kernel invocation or are in kernels invoked with the same launch configuration (i.e. `range` or `nd_range`); and
> * The same values are contributed to each reduction; and
> * The work-items in each reduction contribute those values in the same pattern and the same order.  For example, if the first reduction contributes values _V1_, _V2_, and _V3_ from one work-item (in that order); then the second reduction must also contribute values _V1_, _V2_, and _V3_ from a work-item (in that order).

I added the part about "same order" here, which I assume is also a requirement?

BTW, I'm surprised this isn't always the case anyway.  What is an example of a case when the implementation wouldn't provide this guarantee even without this property?

Reply:
I changed "calls" to "reduction" and removed the "in the same kernel invocation" part, otherwise made the change to the suggested text in a33abc41b30848b234bf7baf4af09291b4b8d864.  I think this is much clearer, so thanks for your help in working through this!

In practice, many reduction implementations trade away determinism for simplicity of implementation and/or improved performance.  The simplest way to implement `sycl::reduction` is to immediately and atomically combine every intermediate result with the final memory location.  A slightly smarter way is to aggregate intermediate results within the sub-group and/or work-group, then atomically combine the results from each group with atomics.  But any atomic combination steps are racy -- so the order of combination depends on which work-items/groups get there first -- and any aggregation within a group makes the order of combination dependent on the launch configuration.  For types/operators that aren't _truly_ commutative and associative (e.g., IEEE floating-point addition) the potential re-ordering makes the result non-deterministic.

Reply:
So, the order of the atomic operations also needs to be the same?  Does this mean that the atomic operations need to be in some strict order according the the ID of the work-item?  If that is the case, maybe the conditions need to be even more strict.  For example, let's say values _V1_, _V2_, and _V3_ are contributed by work-item ID 0 in one reduction.  Do those values also need to be contributed by work-item ID 0 in the second reduction?

Reply:
Yes, you're right.  Clarified further in 05d0cb4eba37882759afa17a50831f1f675197dc.